orc/src/main/java/org/apache/iceberg/orc/OrcMetrics.java

```diff
@@ -0,0 +1,154 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.iceberg.Metrics;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+import org.apache.iceberg.hadoop.HadoopInputFile;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.types.Conversions;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+import org.apache.orc.ColumnStatistics;
+import org.apache.orc.OrcFile;
+import org.apache.orc.OrcProto;
+import org.apache.orc.Reader;
+import org.apache.orc.TypeDescription;
+
+public class OrcMetrics {
+
+  private OrcMetrics() {}
+
+  public static Metrics fromInputFile(InputFile file) {
+    final Configuration config = (file instanceof HadoopInputFile)
+        ? ((HadoopInputFile)file).getConf()
+        : new Configuration();
+    return fromInputFile(file, config);
+  }
+
+  public static Metrics fromInputFile(InputFile file, Configuration config) {
+    try {
+      final Reader orcReader = OrcFile.createReader(new Path(file.location()),
+          OrcFile.readerOptions(config));
+      final TypeDescription orcSchema = orcReader.getSchema();
+      final Schema schema = TypeConversion.fromOrc(orcSchema);
+
+      ColumnStatistics[] colStats = orcReader.getStatistics();
+      List<OrcProto.ColumnStatistics> colStatsProto = orcReader.getOrcProtoFileStatistics();
+      Map<Integer, Long> columSizes = Maps.newHashMapWithExpectedSize(colStats.length);
+      Map<Integer, Long> valueCounts = Maps.newHashMapWithExpectedSize(colStats.length);
+      Map<Integer, ByteBuffer> lowerBounds = Maps.newHashMap();
+      Map<Integer, ByteBuffer> upperBounds = Maps.newHashMap();
+
+      for (int i = 0; i < colStats.length; i++) {
+        columSizes.put(i, colStats[i].getBytesOnDisk());
+        valueCounts.put(i, colStats[i].getNumberOfValues());
+
+        final OrcProto.ColumnStatistics protoStats = colStatsProto.get(i);
+        final Types.NestedField col = schema.findField(i);
+        if (col != null) {
+          Optional<ByteBuffer> orcMin = fromOrcMin(col, protoStats);
+          if (orcMin.isPresent()) {
+            lowerBounds.put(i, orcMin.get());
+          }
+          Optional<ByteBuffer> orcMax = fromOrcMax(col, protoStats);
+          if (orcMax.isPresent()) {
+            upperBounds.put(i, orcMax.get());
+          }
+        }
+      }
+
+      return new Metrics(orcReader.getNumberOfRows(),
+          columSizes,
+          valueCounts,
+          Maps.newHashMap(),
+          lowerBounds,
+          upperBounds);
+    } catch (IOException ioe) {
+      throw new RuntimeIOException(ioe, "Failed to read footer of file: %s", file);
+    }
+  }
+
+  private static Optional<ByteBuffer> fromOrcMin(Types.NestedField column,
+                                                    OrcProto.ColumnStatistics columnStats) {
```

Comment:
is OrcProto.ColumnStatistics a pubic API for ORC . I see that proper column statistics interfaces are also defined e.g see [DecimalColumnStatistics](https://github.com/apache/orc/blob/master/java/core/src/java/org/apache/orc/DecimalColumnStatistics.java).

I guess the Proto classes are used to avoid a typecast?

Reply:
Found the right interfaces! Thanks for the pointers.